docs/enhancements/add-external-providers.md

```diff
@@ -0,0 +1,187 @@
+ 
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a RHV admin I want to create an Openshift cluster on my RHV engine by using the assisted installer.
+
+#### Story 2
+
+As a RHV admin I want to create an Openshift cluster on my RHV engine by using the assisted installer in which
+Control-Plain (Master) nodes are deployed on top of RHV and Compute (Worker) nodes are Bare-Metal nodes.
+
+#### Story 3
+
+As an Openshift developer working on platform X I want to add support for installing Openshift with assisted installer
+on top of platform X.
+
+### Implementation Details/Notes/Constraints
+
+#### Changes to the assisted-service project:
+
+##### Step 1: Detect platform specific logic
+
+Go over the assisted installer flow and detect parts of the code which require provider specific logic.
+
+##### Step 2: Create a Go interface
+
+We should create a Go interface that will contain each provider specific logic in a function
+for each provider to implement.
+
+For example, prototype:
+
+```go
+// Provider contains functions which are required to support installing on a specific platform.
+type Provider interface {
+    // Name returns the name of the platform.
+    Name() string
+    // UpdatePlatformInventoryData sets all platform specific inventory data on the updates map.
+    UpdatePlatformInventoryData(params installer.UpdateClusterParams, cluster *common.Cluster, updates map[string]interface{}) error
+    // SetPlatformInstallConfig sets all platform specific install config parameters
+    SetPlatformInstallcfgValues(platform *platformVsphere, clusterPlatform *models.VspherePlatform)
+    //...
+}
+```
+
+##### Step 3: Implement registry
+
+Implement the registry pattern so providers can register with their names.
+When we will detect the a certain provider then we will register it.
+If no provider is detected then we register the Bare-Metal provider.
+
+##### Step 4: Implement providerDetector method
+
+Implement a function which determines which providers can be used.
+Each provider should implement its own detection logic.
+For example:
+if all control plain nodes are running on oVirt
+(i.e have [SystemVendor.Manufacturer](https://github.com/openshift/assisted-service/blob/master/models/system_vendor.go)
+== oVirt OR RHV) then register the oVirt provider.
+
+##### Step 5: Add calls to the provider functions
+
+Whenever we need to call some provider-specific function we should use the provider interface instead of a switch case.
+
+##### Step 6: Implement Default Provider
+
+Will contain default implementation of most functions required by Provider interface. 
+It should be embedded in all providers to make adding new APIs easier.
+
+##### Step 7: Implement the oVirt provider and migrate Baremetal and vSphere providers
+
+Implement the oVirt provider and migrate existing providers to the Provider interface.
+
+#### Changes to the assisted-agent project:
```

Comment:
We need another interface for validations .. some will be per provider

Reply:
Yes you are right, I'll add flows added on https://github.com/openshift/assisted-service/pull/2274